common/components/sorting.hpp.inc

```diff
@@ -0,0 +1,314 @@
+            bitonic_cas(els[i], other, new_reverse);
+        }
+        half::merge(els, reverse);
+    }
+
+    __device__ static void sort(ValueType *els, bool reverse)
+    {
+        auto new_reverse = reverse != upper_half();
+        half::sort(els, new_reverse);
+        merge(els, reverse);
+    }
+};
+
+template <typename ValueType, int NumLocalElements>
+struct bitonic_warp<ValueType, NumLocalElements, 1> {
+    using local = bitonic_local<ValueType, NumLocalElements>;
+    __device__ static void merge(ValueType *els, bool reverse)
+    {
+        local::merge(els, reverse);
+    }
+    __device__ static void sort(ValueType *els, bool reverse)
+    {
+        local::sort(els, reverse);
+    }
+};
+
+
+/**
+ * @internal
+ * This is a recursive implementation of a bitonic sorting network,
+ * executed in parallel in a thread block using shared memory.
+ *
+ * We use a tiled storage pattern to avoid memory bank collisions on shared
+ * memory accesses, see @ref shared_idx.
+ */
+template <typename ValueType, int num_local, int num_threads, int num_groups,
+          int num_total_threads>
+struct bitonic_global {
+    constexpr static auto num_elements = num_local * num_threads * num_groups;
+    using half = bitonic_global<ValueType, num_local, num_threads,
+                                num_groups / 2, num_total_threads>;
+    static_assert(num_groups > 0, "number of groups must be positive");
+    static_assert(num_threads > 0,
+                  "number of threads per group must be positive");
+    static_assert(num_local > 0, "number of local elements must be positive");
+    static_assert(num_total_threads > 0, "number of threads must be positive");
+    static_assert(32 % num_groups == 0,
+                  "num_groups must be a power of two <= 32");
+
+    __device__ static int shared_idx(int local)
+    {
+        auto rank = group::this_thread_block().thread_rank();
+        // use the same memory-bank to avoid bank conflicts
+        return rank + local * num_total_threads;
+    }
+
+    // check if we are in the upper half of all groups in this block
+    // this is important as for sorting, we have to reverse the sort order in
+    // the upper half
+    __device__ static bool upper_half()
+    {
+        auto rank = group::this_thread_block().thread_rank();
+        return bool(rank & (num_groups * num_threads / 2));
+    }
+
+    __device__ static void merge(ValueType *local_els, ValueType *shared_els,
+                                 bool reverse)
+    {
+        group::this_thread_block().sync();
+        auto upper_shared_els = shared_els + (num_groups * num_threads / 2);
+        // only the lower group executes the CAS
+        if (!upper_half()) {
+            for (auto i = 0; i < num_local; ++i) {
+                auto j = shared_idx(i);
+                bitonic_cas(shared_els[j], upper_shared_els[j], reverse);
+            }
+        }
+        half::merge(local_els, shared_els, reverse);
```

Comment:
After merge, does it also need sync?
bitonic_global<ValueType, num_local, num_threads, 1, num_total_threads>::merge operates on the shared memory

Reply:
I don't think this is necessary, as I assumed as a pre- and postcondition for all `merge` and `sort` functions that the threads are not synchronized with respect to shared memory, and thus only read from shared memory after a sync. For your concrete question, see L237. Does that make sense to you?

Reply:
After the bitonic sort call, we need sync to make sure the operation is done, is that correct?

Reply:
I believe not, since the results are stored in local variables in the last stages, not shared memory. I also specified in the comments that the shared-memory buffer does not use a contiguous layout, but is striped to avoid memory bank collisions. So ideally you should only use the local variables before and after the bitonic sorting function.

Reply:
got it. thanks